add page on validator keys from kb.beaconchain (#7215)
ethereum/ethereum-org-website: src/content/developers/docs/consensus-mechanisms/pos/keys/index.md

---

> title: Keys in proof-of-stake Ethereum

Reviewer: Same comment as below. Any reason this can't just be Keys on Ethereum?

Reviewer 2: I think including proof-of-stake makes sense since this page focuses on validator keys. With "Keys on Ethereum" I would assume this to focus more on regular wallet keys. There is a ton of overlap in the topics though, which is why I suggested we potentially expand on this in the future with more content about keys, but this PR is under `consensus-mechanisms/pos/` and I think this makes sense as is.

Reviewer: Would it be correct just to call them validator keys?

Author: eh - could do, but withdrawal keys are also included on the page.

Reviewer 2: Yeah this is a little tricky.
I tend to think of the content here slightly differently, with "validator keys" consisting of both "signing keys" and "withdrawal keys". The signing keys are "hot" and stay within or available to your validator client.

Only nuance there, is _all_ keys can "sign", so it kinda adds a different naming conflict when trying to explain this stuff.

@jmcook1186 Am I incorrect in this line of thought? Would it be improper to change the "The validator key" section below to "The signing key", and use that terminology instead?

We can then refer to the whole set as "validator keys".

---

Suggestion:
> `title: Keys in proof-of-stake Ethereum`

`title: "Validator keys"`

> `also required a **validator key** and a **withdrawal key**.`

`also required a validator **signing key** and a **withdrawal key**.`

> `### The validator key `

`### The signing key`

And also updating other instances of "validator" key to "signing" where appropriate, or optionally "validator signing keys" occasionally.

Author: idk i don't love this bc the withdrawal key is not functionally different from the validator key. "Validator" and "withdrawal" keys separate them by their intended uses - i.e. validator keys are used for validator duties, withdrawal kays are used to withdraw stake - "signing  keys" suggests the alternative is "non-signing keys" which is not true.

Author: it's also consistent with previous docs, so I say keep validator/withdrawal for now and update later if necessary?
@minimalsm

> description: An explanation of keys used in Ethereum since the transition to proof-of-stake.

Reviewer: I think we can explain (most) things as how they are now (post-merge) vs relative to how they were.

```suggestion
description: An overview of keys used in Ethereum.
```

Reviewer 2: Similar to above, I think keeping it dialed into PoS is appropriately here.
```suggestion
description: An explanation of keys used in Ethereum's proof-of-stake consensus mechanism
```

lang: en
sidebar: true
---

Ethereum secures user's assets using public-private key cryptography. The public key is used as the basis for an Ethereum address - that is, it is visible to the general public and used as a unique identifier. The private (or 'secret') key should only ever be accessible to an account owner. The private key is used to 'sign' transactions and data so that cryptography can prove that the holder approves some action of a specific private key.

Ethereum's keys are generated using [elliptic-curve cryptography](https://en.wikipedia.org/wiki/Elliptic-curve_cryptography).

> However, when Ethereum switched from [proof-of-work](/developers/docs/consensus-mechanisms/pow) to [proof-of-stake](/developers/docs/consensus-mechanisms/pos) a new type of key was added to Ethereum. The original keys still work exactly the same as before - there were no changes to the elliptic-curve-based keys securing accounts. However, users needed a new type of key for participating in proof-of-stake by staking ETH and running validators. This need arose from scalability challenges associated with many messages passing between large numbers of validators that required a cryptographic method that could easily be aggregated to reduce the amount of communication required for the network to come to consensus.

Reviewer: Agree. I don't love this stylistically either. I'd maybe drop the first use of 'however' in this paragraph altogether, as it's a little monotonous and confusing.

This new type of key uses the [Boneh-Lyn-Shacham (BLS) signature scheme](https://wikipedia.org/wiki/BLS_digital_signature). BLS enables a very efficient aggregation of signatures but also allows reverse engineering of aggregated individual validator keys and is ideal for managing actions between validators.

## The two types of validator key {#two-types-of-key}

Before the switch to proof-of-stake, Ethereum users only had a single elliptic-curve-based private key to access their funds. With the introduction of proof-of-stake, users that wished to be solo stakers also required a **validator key** and a **withdrawal key**.

### The validator key {#validator-key}

The validator signing key consists of two elements:

- Validator **private** key
- Validator **public** key

> The purpose of the validator private key is to actively sign on-chain operations such as block proposals and attestations. Therefore these keys have to be held in a hot wallet.

Reviewer: ```suggestion
The purpose of the validator private key is to sign on-chain operations such as block proposals and attestations. Because of this, these keys must be held in a hot wallet.
```

> This flexibility has the advantage to move validator signing keys very quickly from one device to another, however, if they have gotten lost or stolen, a thief may be able to **act maliciously** in the following ways:

Reviewer: ```suggestion
This flexibility has the advantage of moving validator signing keys very quickly from one device to another. However, if they have gotten lost or stolen, a thief may be able to **act maliciously** in several ways:
```

Reviewer: > English question, wouldn't wrapping "however" in commas be the more appropriate way to use it here? Always tried to avoid using it to start sentences where possible.

I changed it to a new sentence because the use of 'however' created a [run-on sentence](https://www.grammarly.com/blog/run-on-sentence-basics/). Using a semi-colon instead would also be correct.

re: your aversion to beginning sentences with 'However', I've not really heard of that before 😛 Maybe a British/American English difference?

> Also, not sure I'd switch to "several" here... It's really just two things, they could get you slashed or they could exit the validator.

I'm good with using 'in a few ways' here. My edit was to remove the implication that the list was a complete list of potentially malicious activities.

- Get the validator slashed) by:
  - Being a proposer and signing two different beacon blocks for the same slot
  - Being an attester and signing an attestation that "surrounds" another one.
  - Being an attester and signing two different attestations having the same target.
- Force a voluntary exit, which stops the validator from staking, and grants access to its ETH balance to the withdrawal key owner.

> The **validator public key** is included in the _deposit data_ which allows Ethereum to identify the validator.

Reviewer: ```suggestion
The **validator public key** is included in the _deposit data_, which allows Ethereum to identify the validator.
```

### The withdrawal key {#withdrawal-key}

The withdrawal key will be required to move the validator balance after this is enabled in the upcoming Shanghai upgrade. Just like the validator keys, the withdrawal keys also consist of two components:

- Withdrawal **private** key
- Withdrawal **public** key

> Losing this key means losing access to the validator balance. However, the validator can still sign attestations and blocks since these actions require the validator private key, but there is little to no incentive to do so if the keys are lost.

Reviewer: ```suggestion
Losing this key means losing access to the validator balance. However, the validator can still sign attestations and blocks since these actions require the validator's private key, but there is little to no incentive if the keys are lost.
```

> Separating out the validator keys from the Ethereum account keys enables multiple validators to be run by a single user.

Reviewer: ```suggestion
Separating the validator keys from the Ethereum account keys enables multiple validators to be run by a single user.
```

> ![validator key schematic](validator-key-schematic.png)

Reviewer 2: ❤️

## Deriving keys from other keys {#deriving-keys-from-keys}

> If every 32 ETH staked required a new set of 2 completely independent keys to be created, key management would quickly become unwieldy especially for users running multiple validators. To get around this problem, multiple validator keys can be derived from a single common secret. Storing that single secret allows access to multiple validator keys.

Reviewer: ```suggestion
If every 32 ETH staked required a new set of 2 completely independent keys, key management would quickly become unwieldy, especially for users running multiple validators. Instead, multiple validator keys can be derived from a single common secret and storing that single secret allows access to multiple validator keys.
```

> [Mnemonics](https://en.bitcoinwiki.org/wiki/Mnemonic_phrase) and paths are a well-known feature that users often encounter when [they access](https://ethereum.stackexchange.com/questions/19055/what-is-the-difference-between-m-44-60-0-0-and-m-44-60-0) their hardware wallets. The mnemonic is a sequence of words that act as an initial seed for a private key. When combined with additional data, the mnemonic is used to generate a hash known as the 'master key'. This can be thought of as the root of a tree. Branches from this root can then be drived using a heirarchical path, so that child nodes can exist as combinations of their parent node's hash and their index in the tree.

Reviewer: ```suggestion
[Mnemonics](https://en.bitcoinwiki.org/wiki/Mnemonic_phrase) and paths are prominent features that users often encounter when [they access](https://ethereum.stackexchange.com/questions/19055/what-is-the-difference-between-m-44-60-0-0-and-m-44-60-0) their hardware wallets. The mnemonic is a sequence of words that act as an initial seed for a private key. When combined with additional data, the mnemonic generates a hash known as the 'master key'. This can be thought of as the root of a tree. Branches from this root can then be derived using a hierarchical path so that child nodes can exist as combinations of their parent node's hash and their index in the tree.
```

Author: ```suggestion
[Mnemonics](https://en.bitcoinwiki.org/wiki/Mnemonic_phrase) and paths are prominent features that users often encounter when [they access](https://ethereum.stackexchange.com/questions/19055/what-is-the-difference-between-m-44-60-0-0-and-m-44-60-0) their wallets. The mnemonic is a sequence of words that act as an initial seed for a private key. When combined with additional data, the mnemonic generates a hash known as the 'master key'. This can be thought of as the root of a tree. Branches from this root can then be derived using a hierarchical path so that child nodes can exist as combinations of their parent node's hash and their index in the tree. Read about [BIP-32](https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki) and [BIP-19](https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki) standards for mnemonic-based key generation.
```

> These paths have the following structure which will be familiar to users who have interacted with hardware wallets:

Reviewer: ```suggestion
These paths have the following structure, which will be familiar to users who have interacted with hardware wallets:
```

```
m/44'/60'/0'/0`
```

The slashes in this path separate components of the private key as follows:

```
master_key / purpose / coin_type / account / change / address_index
```

> This logic enables a user to attach as many validators to a single **withdrawal key** as they want, because the root of the tree can be common and differentiation can happen at the branches. The user can **derive any number of keys** from the Mnemonic phrase.

Reviewer: ```suggestion
This logic enables users to attach as many validators to a single **withdrawal key** as possible because the tree root can be common, and differentiation can happen at the branches. The user can **derive any number of keys** from the Mnemonic phrase.
```

Reviewer 2: Agree with Josh's suggestions, but also, I'm not sure I understand the way this is worded.

My understanding is each validator has its own set of priv/pub signing keys + priv/pub withdrawal keys. For example, two validators would need two different withdrawal keys to eventually withdraw funds. Both of these can be derived from the seed, but are still unique.

But, with how it's worded here, seems like you're saying a single withdrawal key can be shared... I assume you're referring to a key higher up in the tree, but this threw me off a little. May be nitpicking, but should we maybe refer to this as the "mnemonic phrase" to be consistent with the image?

```suggestion
This logic enables users to attach as many validators as possible to a single **mnemonic phrase** because the tree root can be common, and differentiation can happen at the branches. The user can **derive any number of keys** from the mnemonic phrase.
```

These paths have the following structure which will be familiar to users who have interacted with hardware wallets:

```
m/44'/60'/0'/0`
```

The slashes in this path separate components of the private key as follows:

```
master_key / purpose / coin_type / account / change / address_index
```

This logic enables a user to attach as many validators to a single **withdrawal key** as they want, because the root of the tree can be common and differentiation can happen at the branches. The user can **derive any number of keys** from the Mnemonic phrase.

```
      [m / 0]
     /
    /
[m] - [m / 1]
    \
     \
      [m / 2]
```

Each branch is separated by a `/` so `m/2` means start with the master key and follow branch 2. In the schematic below a single mnemonic phrase is used to store three withdrawal keys, each with two associated validators.

![validator key logic](multiple-keys.png)

## Further Reading {#further-reading}

> [Ethereum Foundation blog post by Carl Beekhuizen](https://blog.ethereum.org/2020/05/21/keys/)
> 
> [EIP-2333 BLS12-381 key generation](https://eips.ethereum.org/EIPS/eip-2333)

Reviewer: Further reading should always be in a markdown list

```suggestion
- [Ethereum Foundation blog post by Carl Beekhuizen](https://blog.ethereum.org/2020/05/21/keys/)
- [EIP-2333 BLS12-381 key generation](https://eips.ethereum.org/EIPS/eip-2333)
```
